### PR TITLE
Add logits method to Sequence class

### PR DESCRIPTION
**Pull request by SamDuffield**
Allows the user to access the next token logits for a given prompt (including with regex!)

```python
prompt = "What is the IP address of the Google DNS servers? "
unguided_logits = generate.continuation(model, max_tokens=30).logits(prompt)
guided_logits = generate.regex(
    model,
    r"((25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}(25[0-5]|2[0-4]\d|[01]?\d\d?)",
    max_tokens=30,
).logits(prompt)
```

### CODE REVIEW

**Comment by rlouf** [comment]
I think it would make more sense to return the logprob as part of the `step` method?

Then we could create a custom object that contains both the text output and the logprob. This might do for now, give me some time to think about it.

What are you using this for?

**Comment by SamDuffield** [comment]
> I think it would make more sense to return the logprob as part of the `step` method?
> 
> Then we could create a custom object that contains both the text output and the logprob. This might do for now, give me some time to think about it.

If you include it in the `step` are you thinking to return the logit only for the sampled token or all of the logits. I think it would be useful for the user (at least me 😄 ) to have access to all of the logits if needed.


> What are you using this for?

It's just very useful to be able to analyse the next token logits to e.g. assess the confidence of the model on a True/False question for various questions (and this pairs very well with regex).

**Comment by SamDuffield** [comment]
Ah I see what you mean, `step` already outputs the `probs` as an additional output.

It would be useful to have an additional method with a simple API like
`generate.continuation(model).next_token_logits(prompt)`
or
`generate.continuation(model).next_token_probs(prompt)`

I.e. returns the next token logits or probs without actually doing any sampling.

**Comment by SamDuffield** [comment]
Ok I've thought about this a bit more.

IMO the best API would be to have a `.next_token_logits(prompt)` or `.next_token_probs(prompt)` function that allows the user to examine the distribution without sampling. I think it's good to have this separate from the `step` function which samples and appends the new token, which is a different use case.

With this new function it may also make sense to have `step` only return the `token_ids` and not `probs`. This could also avoid some confusion as the `token_ids` relate to the whole sequence and `probs` only the appended token.

Was there a specific reason to have `step` return `probs` as well?

WDYT? @rlouf @brandonwillard

**Comment by rlouf** [comment]
`next_token_logits` sounds good

**Comment by SamDuffield** [comment]
I'm now thinking that indeed the `probs` might be useful for SMC 😃 

Let's leave `step` returning `probs` for now

File: outlines/text/generate/sequence.py [hunk 1]
```diff
@@ -40,6 +40,38 @@ def is_finished(self, token_ids: torch.LongTensor) -> torch.BoolTensor:
...
+            The input prompt.
+
+        Returns
+        -------
+        An array of shape `(batch_size, vocab_size)` containing the logits
+        (unnormalised log probabilities) for the next token generation.
+
+        """
+        token_ids, attention_mask = self.model.tokenizer.encode(prompt)
+        num_prompt_tokens = token_ids.shape[-1]
```
Comment (SamDuffield) [inline outlines/text/generate/sequence.py:68]: I'm wondering if this is working as intended in the vectorised case.

Surely in the vectorised case, different prompts will have different lengths

Comment (SamDuffield) [reply outlines/text/generate/sequence.py:68]: This was my bad, indeed the attention_mask handles it.

I was getting confused seeing different probs on different runs but it was because I was using GPT2 which has Dropout

**Comment by dpsimpson** [comment]
(Also sorry @rlouf I must've mis-clicked. I don't think I meant to remove you as a reviewer)
